ksqldb-engine/src/main/java/io/confluent/ksql/serde/SerdeFeaturesFactory.java

```diff
@@ -112,34 +129,33 @@ public static SerdeFeatures buildValueFeatures(
    * classes to ensure that anytime a key is changed we properly set the key format.
    *
    * @param keyFormat un-sanitized format
-   * @param numKeyColumns number of key columns for this stream/table
+   * @param newKeyColumnSqlTypes sql types of key columns for this stream/table
    * @param allowKeyFormatChangeToSupportMultipleKeys safeguard to prevent changing key formats in
    *                                                  unexpected ways. if false, no format change
    *                                                  will take place
    * @return the new key format to use
    */
   public static KeyFormat sanitizeKeyFormat(
       final KeyFormat keyFormat,
-      final int numKeyColumns,
+      final List<SqlType> newKeyColumnSqlTypes,
       final boolean allowKeyFormatChangeToSupportMultipleKeys
   ) {
     return sanitizeKeyFormatWrapping(
         sanitizeKeyFormatForMultipleColumns(
             keyFormat,
-            numKeyColumns,
+            newKeyColumnSqlTypes,
```

Comment:
nit: can we add a third type of sanitization, rather than combining the logic for multiple key columns and specific key column types into a single method? The two are logically different and can be handled independently from each other. I think this will make the logic in this file easier to read.